src/precompile.c

```diff
@@ -125,38 +125,52 @@ JL_DLLEXPORT void jl_write_compiler_output(void)
 
     // jl_dump_native writes the clone_targets into `s`
     // We need to postpone the srctext writing after that.
-    if (native_code) {
+    // Note that this region of code is carefully ordered to minimize
+    // live memory at any given point. Consider this when making changes.
+    if (emit_native) {
+        assert(native_code && "Forgot to emit native code?");
+        ios_t f;
+        if (emit_split) {
+            if (ios_file(&f, jl_options.outputji, 1, 1, 1, 1) == NULL)
+                jl_errorf("cannot open system image file \"%s\" for writing", jl_options.outputo);
+            ios_write(&f, (const char*)s->buf, (size_t)s->size);
+            ios_close(s);
+            free(s);
+        }
+        void *sysimg_data_module = jl_create_sysimg_data_module(native_code, (const char*)z->buf, (size_t)z->size);
```

Comment:
I think we could save a good amount of additional memory if we move all of this, along with the  jl_create_system_image call inside the jl_dump_native function (just before we need it).